File: PCA.R
data.matrix <- matrix(nrow=100, ncol=10)
colnames(data.matrix) <- c(
  paste("wt", 1:5, sep=""), # wild type: normal, every day samples
  paste("ko", 1:5, sep="") # knock-out: samples missing a gene because we knocked it out
)
rownames(data.matrix) <- paste("gene", 1:100, sep="")
for (i in 1:100) {
  wt.values <- rpois(5, lambda=sample(x=10:1000, size=1)) # Poisson distribution
  ko.values <- rpois(5, lambda=sample(x=10:1000, size=1))

  data.matrix[i,] <- c(wt.values, ko.values)
}
head(data.matrix)

pca <- prcomp(t(data.matrix), scale=TRUE)
# By default, prcomp() expects the samples to be rows and the genes to be columns.
# Thus, transpose the data.matrix.
# If not, we will get a graph that showss how the genes are related to each other.
# prcomp() returns:
# 1. x: the PCs for drawing a graph
# 2. sdev
# 3. rotation

# Use the first two columns in x to draw a 2-D plot that uses the first two PCs.
plot(pca$x[,1], pca$x[,2]) # PC1, PC2
# The first PC accounts for the most variation in the original data

# calculate how much variation in the original data each PC accounts for
pca.var <- pca$sdev^2
# the percentage of variation that each PC accounts for
pca.var.per <- round(pca.var/sum(pca.var)*100, 1)
barplot(pca.var.per, main="Scree Plot", xlab="Principal Component",
  ylab="Percent Variation")
# There's a big difference between the two clusters in plot(pca$x[,1], pca$x[,2])

library(ggplot2)

pca.data <- data.frame(Sample=rownames(pca$x),
  X=pca$x[,1],
  Y=pca$x[,2])
pca.data

ggplot(data=pca.data, aes(x=X, y=Y, label=Sample)) +
  geom_text() + # plot the labels, rather than dots or other shapes
  xlab(paste("PC1 - ", pca.var.per[1], "%", sep="")) +
  ylab(paste("PC2 - ", pca.var.per[2], "%", sep="")) +
  theme_bw() + # makes the graph's background white
  ggtitle("My PCA Graph")

# Look at the loading scores for PC1 only
# since it accounts for 91.5% of the variation in the data
loading_scores <- pca$rotation[,1]
gene_scores <- abs(loading_scores)
# Genes that push samples to the left side of the graph have large -ve values,
# genes that push samples to the right have large +ve values.
# Use abs() to sort based on the numbers' magnitude rather than from high to low
gene_score_ranked <- sort(gene_scores, decreasing=TRUE)
top_10_genes <- names(gene_score_ranked[1:10])
top_10_genes

pca$rotation[top_10_genes, 1] # show the scores and signs

File: PCA.py
import pandas as pd # pandas: panel data
import numpy as np
import random as rd
from sklearn.decomposition import PCA
from sklearn import preprocessing # provides function for scaling the data before PCA
from sklearn.preprocessing import StandardScaler
import matplotlib.pyplot as plt

genes = ['gene' + str(i) for i in range(1,101)]
wt = ['wt' + str(i) for i in range(1,6)]
ko = ['ko' + str(i) for i in range(1,6)]
data = pd.DataFrame(columns=[*wt, *ko], index=genes)
# * unpacks the "wt" and "ko" arrays so that the column names are a single array:
# [wt1, wt2, ..., ko1, ko2, ..., ko5]
# Without *, it will be an array of two arrays:
# [[wt1, ..., wt5], [ko1, ..., ko5]]

# For each gene in the index (gene1, ..., gene100),
# create 5 values for the "wt" samples nad 5 values for the "ko" samples.
for gene in data.index:
    data.loc[gene,'wt1':'wt5'] = np.random.poisson(lam=rd.randrange(10,1000), size=5)
    data.loc[gene,'ko1':'ko5'] = np.random.poisson(lam=rd.randrange(10,1000), size=5)
    # For each gene, select a new mean for the poisson distribution, vary between 10 and 1000.

print(data.head())
print(data.shape) # returns the dimensions of the data matrix eg. (100, 10)

# centre and scale the data
scaled_data = preprocessing.scale(data.T) # If it is columns, tranpose it before analysis
# After centering, the average value for each gene will be 0.
# After scaling, the sd for the values for each gene will be 1.

# Alternatively, more commonly used for machine learning,
scaled_data = StandardScaler().fit_transform(data.T)

# In sklearn, variation is calculated with denominator equals n.
# In R using scale() or prcomp(), the denominator is n - 1 instead.
# The latter results in larger, but unbiased, estimates of the variation.
# These differences don't affect the PCA analysis.
# The loading scores and the amount of variation per PC will be the same.
# However, these differences will have a minor effect to the final graph.
# The coordinates on the final graph come from multiplying the loading scores by the scaled values.
pca = PCA() # machine learning step
pca.fit(scaled_data) # calculate loading scores and the variation each PC accounts for
pca_data = pca.transform(scaled_data) # generate coordinates for a PCA graph based on the loading scores and the scaled data.

# calculate the percentage of variance each PC accounts for
per_var = np.round(pca.explained_variance_ratio_* 100, decimals=1)
labels = ['PC' + str(x) for x in range(1, len(per_var)+1)]

plt.bar(x=range(1, len(per_var)+1), height=per_var, tick_label=labels)
plt.ylabel('Percentage of Explained Variance')
plt.xlabel('Principal Component')
plt.title('Scree Plot')
plt.show()

# put the new coordiantes (pca_data) into a matrix where the rows have sample labels and columns have PC labels
pca_df = pd.DataFrame(pca_data, index=[*wt, *ko], columns=labels)

plt.scatter(pca_df.PC1, pca_df.PC2)
plt.title('My PCA Graph')
plt.xlabel('PC1 - {0}%'.format(per_var[0]))
plt.ylabel('PC2 - {0}%'.format(per_var[1]))
# add sample names to the graph
for sample in pca_df.index:
    plt.annotate(sample, (pca_df.PC1.loc[sample], pca_df.PC2.loc[sample]))
plt.show()
